consensus/core/src/commit_syncer.rs

```diff
@@ -0,0 +1,586 @@
+                Some(result) = inflight_fetches.join_next(), if !inflight_fetches.is_empty() => {
+                    if let Err(e) = result {
+                        warn!("Fetch cancelled or panicked, CommitSyncer shutting down: {}", e);
+                        // If any fetch is cancelled or panicked, try to shutdown and exit the loop.
+                        inflight_fetches.shutdown().await;
+                        return;
+                    }
+                    let (target_end, commits, blocks): (CommitIndex, Vec<TrustedCommit>, Vec<VerifiedBlock>) = result.unwrap();
+                    assert!(!commits.is_empty());
+                    let metrics = &inner.context.metrics.node_metrics;
+                    metrics.commit_sync_fetched_commits.inc_by(commits.len() as u64);
+                    metrics.commit_sync_fetched_blocks.inc_by(blocks.len() as u64);
+
+                    let (commit_start, commit_end) = (commits.first().unwrap().index(), commits.last().unwrap().index());
+                    // Allow returning partial results, and try fetching the rest separately.
+                    if commit_end < target_end {
+                        pending_fetches.push_front((commit_end + 1, target_end));
+                    }
+                    // Make sure synced_commit_index is up to date.
+                    synced_commit_index = synced_commit_index.max(inner.dag_state.read().last_commit_index());
+                    // Only add new blocks if at least some of them are not already committed.
+                    if synced_commit_index < commit_end {
+                        fetched_blocks.insert((commit_start, commit_end), blocks);
+                    }
+                    // Try to process as many blocks as possible, as long as there is no gap between
+                    'fetched: while let Some(((fetched_start, _end), _blocks)) = fetched_blocks.first_key_value() {
+                        // Only pop fetched_blocks if there is no gap with blocks already sent to Core.
+                        // Note: start, end and synced_commit_index are all inclusive.
+                        let ((_start, fetched_end), blocks) = if *fetched_start <= synced_commit_index + 1 {
+                            fetched_blocks.pop_first().unwrap()
+                        } else {
+                            // Found gap between earliest fetched block and latest synced block,
+                            // so not sending additional blocks to Core.
+                            break 'fetched;
+                        };
+                        // Avoid sending to Core a complete batch of duplicated blocks.
+                        if fetched_end <= synced_commit_index {
+                            continue 'fetched;
+                        }
+                        // If core thread cannot handle the incoming blocks, it is ok to block here.
+                        if let Err(e) = inner.core_thread_dispatcher.add_blocks(blocks).await {
```

Comment:
Yes it is indeed possible to have blocks suspended here right now. When a validator equivocates, some blocks from the validator may never make it into a commit because only blocks higher than the last commit round of the authority get into new commits. But the block can still be an ancestor to other committed blocks. We need to rely on block sync here instead.

In future maybe we can implement a form of GC to avoid suspending blocks in these cases when ancestors cannot be committed. But it will take a bit of thinking.